specs/appwindow-spec.md

```diff
@@ -0,0 +1,93 @@
+Window.AppWindow api
+===
+New api to simplify accessing [appwindow](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/windowing/windowing-overview)
+functionality through WinUI code
+
+# Background
+Xaml has a [Window](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Window) API that
+internally wraps an hwnd. Windows has an [AppWindow](https://docs.microsoft.com/uwp/api/Windows.UI.WindowManagement.AppWindow) class
+that similarly wraps an hwnd in UWP. WinAppSDK has a new [AppWindow](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Windowing.AppWindow) which wraps an hwnd and works on Desktop.
+
+You can get an AppWindow from a Xaml Window by calling a COM API to get Xaml and then
+a DLL export API to convert the hwnd to an AppWindow. This spec adds a simple `Window.AppWindow`
+property to make this much easier and more discoverable.
+
+[Xaml Window](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window) will
+expose AppWindow object directly to app developer through an api. Instead of writing a lot of boiler plate
+code everywhere, app developer can use this api, reducing code bloat, and making appwindow apis easily
+accessible from winui code.
+
+These *before* and *after* C# code examples illustrate how this api simplifies integrating appwindow apis in winui codebase.
+
+### Before
+```c#
+// This is needed to get any Window from inside a Xaml control 
+var xamlWindow = WindowHelper.GetWindowForElement(this); // api to get window object from UIElement (not a real api)  
+
+// unnecessary boiler plate code 
+var windowId = Win32Interop.GetWindowIdFromWindow(WindowNative.GetWindowHandle(xamlWindow)); 
+var appWindow = AppWindow.GetFromWindowId(windowId); 
+
+//calling function foo  (not a real appwindow function)
+appWindow.foo();
+```
+
+### After
+```c#
+// This is needed to get any Window from inside a Xaml control 
+var xamlWindow = WindowHelper.GetWindowForElement(this); // api to get window object from UIElement (not a real api)  
+   
+
+//calling function foo directly
+xamlWindow.AppWindow.foo();  
+```
+Notice how `xamlWindow.AppWindow.foo()` doesn't require additional steps to call the function `foo`.
+
+
+# API Pages
+
+## Window.AppWindow property
+
+Gets the `AppWindow` associated with this `Window`.
+```c#
+public Microsoft.UI.Windowing.AppWindow Window.AppWindow { get; }
+```
+_Spec note: Mapping an hwnd to an AppWindow causes the hwnd to be subclassed, meaning that the timing of
+when [AppWindow.GetFromWindowId](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Windowing.AppWindow.GetFromWindowId)
+is first called can potentially have a behavioral side effect. To ensure this is predictable, the AppWindow
+for Xaml's hwnd will be created during the Xaml Window's construction. Note that this could potentially be
+an observable behavior change from the behavior before the introduction of this API.
```

Comment:
An alternative to the behavior change would be a Get method rather than a getter property. (Method calls are explicit, property getters are less obvious, e.g. can be triggered at seemingly random times by data binding, and can be triggered by putting your mouse in the wrong place in the debugger.)

```
runtimeclass Window
{
    AppWindow GetAppWindow();
}
```

Reply:
Discussion: Note that there's also a performance impact of eagerly creating the AppWindow and causing it to set up its hwnd subclass.

Discussion: There's a lot of value to having predictable ordering of events in terms of the ability to write code and the stability of it. There's some risk to making this behavior change for a shared component that could be running in a processing running either version, and (while the framework is side-by-side in minor releases) when an app retargets to a new version. Even if we did the method approach rather than the property, which would make it technically predictable, it would still appear unpredictable because it's so unobvious.

So the decision is to make the ordering predictable, which is an observable behavior change.